src/http_interface.py

```diff
@@ -0,0 +1,202 @@
+
+logger = logging.getLogger()
+SERVICES_CONFIGURATION_KEY = "services"
+
+
+class HTTPBackendAvailableEvent(RelationEvent):
+    """Event representing that http data has been provided."""
+
+
+class HTTPBackendRemovedEvent(RelationEvent):
+    """Event representing that http data has been removed."""
+
+
+class HTTPProviderEvents(CharmEvents):
+    """Container for HTTP Provider events.
+
+    Attrs:
+        http_backend_available: Custom event when integration data is provided.
+        http_backend_removed: Custom event when integration data is removed.
+    """
+
+    http_backend_available = EventSource(HTTPBackendAvailableEvent)
+    http_backend_removed = EventSource(HTTPBackendRemovedEvent)
+
+
+class _IntegrationInterfaceBaseClass(Object):
+    """Base class for integration interface classes.
+
+    Attrs:
+        relations: The list of Relation instances associated with the charm.
+        bind_address: The unit address.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Initialize the interface base class.
+
+        Args:
+            charm: The charm implementing the requirer or provider.
+            relation_name: Name of the integration using the interface.
+        """
+        super().__init__(charm, relation_name)
+
+        observe = self.framework.observe
+        self.charm: CharmBase = charm
+        self.relation_name = relation_name
+
+        observe(charm.on[relation_name].relation_created, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+        observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_departed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    @abc.abstractmethod
+    def _on_relation_joined(self, _: RelationJoinedEvent) -> None:
+        """Abstract method to handle relation-joined event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _on_relation_changed(self, _: RelationChangedEvent) -> None:
+        """Abstract method to handle relation-changed event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _on_relation_broken(self, _: RelationBrokenEvent) -> None:
+        """Abstract method to handle relation-changed event.
+
+        Raises:
+            NotImplementedError: if the abstract method is not implemented.
+        """
+        raise NotImplementedError
+
+    @property
+    def relations(self) -> list[Relation]:
+        """The list of Relation instances associated with the charm."""
+        relations = self.charm.model.relations.get(self.relation_name, [])
+        return relations
```

Comment:
Any reason for having a variable `relations` here?

Reply:
No particular reason, it was brought over from the traefik lib. I updated the code to return directly